### ISSUE TITLE
Travis Flake in BuildPathTest.testCheckLibraries

### ISSUE THREAD

**Issue by elharo** [GoogleCloudPlatform/google-cloud-eclipse#2758]
WARNING: Project no longer exists: test0.5965265269703218
Tests run: 4, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 3.98 sec <<< FAILURE! - in com.google.cloud.tools.eclipse.appengine.libraries.BuildPathTest
testAddNativeLibrary(com.google.cloud.tools.eclipse.appengine.libraries.BuildPathTest)  Time elapsed: 1.556 sec
testCheckLibraries(com.google.cloud.tools.eclipse.appengine.libraries.BuildPathTest)  Time elapsed: 0.761 sec  <<< FAILURE!
java.lang.AssertionError: null
	at org.junit.Assert.fail(Assert.java:86)
	at org.junit.Assert.assertTrue(Assert.java:41)
	at org.junit.Assert.assertFalse(Assert.java:64)
	at org.junit.Assert.assertFalse(Assert.java:74)
	at com.google.cloud.tools.eclipse.appengine.libraries.BuildPathTest.testCheckLibraries(BuildPathTest.java:110)

**Comment by chanseokoh** [comment]
The test seems flawed. If I suspend the main thread after `project.setRawClasspath()` for a while, it flips the condition that we want to test. The following code fails with an assertion failure 

```java
    assertTrue(project.getProject().exists(librariesIdPath));

    // remove the master-library container, so checkLibraryList should remove the library ids file
    project.setRawClasspath(originalClasspath, null);

    try {
      Thread.sleep(3000);
    } catch (InterruptedException e) {}
    assertTrue(project.getProject().exists(librariesIdPath));  // <--- Fails!

    BuildPath.checkLibraryList(project, null);  // We're tesitng if this call will flip the condition.
                                                // However, `setRawClasspath` seems to do it too.
    assertFalse(project.getProject().exists(librariesIdPath));
```

**Comment by briandealwis** [comment]
I wonder if the container resolving or the callback happens on a different thread…

**Comment by briandealwis** [comment]
Right so in `.libraries`'s `Activator` we install a Java model listener that triggers the `BuildPath.checkLibraryList()` in a `WorkspaceJob`.  So that `Thread.sleep(3000)` provides plenty of time for that job to execute and remove the library file.

`checkLibraryList()` doesn't actually change the classpath itself, it only calls `LibraryClasspathContainerSerializer#removeLibraryIds()` to remove the `_libraries.container` file and `resetContainer()` to remove the `master.container` file.  So calling `checkLibraryList()` in the current thread shouldn't matter.

**Comment by briandealwis** [comment]
The only reason the `_libraries.container` file won't be removed in `BuildPath#checkLibraryList()` is if either (1) the `_libraries.container` file doesn't exist or (2) if our master classpath container is still present.

Or unless the file is somehow recreated…

**Comment by chanseokoh** [comment]
I've long been thinking of the potentially problematic incomplete serialization of `LibraryClasspathContainerSerializer`. For example, when methods like `saveLibraryIds()` write things to a file, they don't do it through the Eclipse `IFile` interface. That is, you actually need to refresh the project to ensure Eclipse can see the changes.

```java
  public void saveLibraryIds(...) ... {
    File librariesFile = getContainerStateFile(javaProject, CONTAINER_LIBRARY_LIST_FILE_ID, true);
    ...
    try (Writer out = Files.newBufferedWriter(librariesFile.toPath(), StandardCharsets.UTF_8)) {
      out.write(gson.toJson(libraryIds.toArray()));
    }
```

**Comment by briandealwis** [comment]
We should change this, but it's complicated as it makes testing hard.  I tried putting a loop around the `BuildPathTest#testCheckLibraries()` to run the test 100 times: it gets to 21 iterations and then fails after the `BuildPath.addNativeLibrary()` call as the `_libraries.container` file doesn't exist.  Maybe we just need to disable the Java model listener during the test.

**Comment by elharo** [comment]
Can we rethink this test in any way so that it does less work?

**Comment by chanseokoh** [comment]
The `Activator.listener` spawns a workspace job that calls `BuildPath.checkLibraryList()`, which is the method of interest in this test. Removing the listener seems to fix it. Running `testCheckLibraries()` thousand times in a loop doesn't seem to fail after this change.

**Comment by chanseokoh** [comment]
Ah, I see the test was written in mind to test just `checkLibraryList`, not the actual interactions by the listener. The listener is the only caller of `checkLibraryList`.

**Comment by chanseokoh** [comment]
The root problem is that the `Activator.listener` (which was to monitor classpath updates to fix #2523) is updating the project classpath and serializing library settings concurrently. I think the best way to handle this concurrency is to not run `Activator.listener` concurrently, i.e., doing things in the main UI thread. I explored if I could synchronize concurrent threads explicitly, but that seems really messy and it will go out of control easily going forward.

**Comment by chanseokoh** [comment]
So, this is not only about fixing the flakiness in tests. I think the flakiness actually revealed a possible race condition due to concurrency that can potentially make project config around our library feature out of proper sync.
